### PR TITLE
wip

### PR DESCRIPTION
**Pull request by nmarton-da**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -69,6 +69,7 @@ start_postgresql
 
 # Run the tests.
 bazel test //... \
+  --cache_test_results=no \
```
Comment (mziolekda) [inline build.sh:72]: This needs to be changed back before it is unwipped

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsPruning.scala [hunk 2]
```diff
@@ -88,7 +88,7 @@ private[backend] trait StorageBackendTestsPruning extends Matchers with StorageB
       contractId = "#1",
       signatory = someParty,
     )
-    val createFilter1 = DbDto.CreateFilter(1L, someTemplateId.toString, "signatory")
+    val createFilter1 = DbDto.CreateFilter(1L, someTemplateId.toString, "party")
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsPruning.scala:91]: Why is this rename required?

**Review by meiersi-da — COMMENTED** [review]
Impressive changes @nmarton-da . Looks like a good direction.

Makes me wonder whether we couldn't implement in-memory fan-out more simply using an aggressively pruned in-memory storage backend for the events that wraps a DB storage backend. For calls that fall within the event-range of the in-memory backend, we'd serve them from memory; the other ones get served from disk.

File: build.sh (hunk 1 above)
Comment (nmarton-da) [reply build.sh:72]: I think we will have other PRs to roll this out, but if no, for sure :)
I added this to trigger full test builds for measurement purposes.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsPruning.scala (hunk 2 above)
Comment (nmarton-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsPruning.scala:91]: This implementation has no filter table, and is getting the filtering info directly from events. So if the test needed to be changed to align the fixtures: same information in create-filters and create-events.
There is a TODO though to add improvement of filter tables (that would be nested maps as indexes in the in memory implementation), because that would be more performant for the in memory as well, and also it would be more conformant as well. (not sure it is needed though)

**Comment by nmarton-da** [comment]
> Impressive changes @nmarton-da . Looks like a good direction.
> 
> Makes me wonder whether we couldn't implement in-memory fan-out more simply using an aggressively pruned in-memory storage backend for the events that wraps a DB storage backend. For calls that fall within the event-range of the in-memory backend, we'd serve them from memory; the other ones get served from disk.

Interesting idea!

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsPruning.scala (hunk 2 above)
Comment (meiersi-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsPruning.scala:91]: Ah OK. Thanks.

File: ledger/participant-integration-api/src/main/scala/platform/store/DbType.scala [hunk 3]
```diff
@@ -43,10 +43,19 @@ object DbType {
         supportsAsynchronousCommits = false,
       )
 
+  object M
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/DbType.scala:46]: M as in Marton or as in M-theory?

```
In the absence of an understanding of the true meaning and structure of M-theory, Witten has suggested that the M should stand for "magic", "mystery", or "membrane" according to taste, and the true meaning of the title should be decided when a more fundamental formulation of the theory is known.
```

**Comment by nmarton-da** [comment]
superseded by associated project
